### PR TITLE
Added additional sensor categories + substance and position properties

### PR DESCRIPTION
**Pull request by hammar**

### CODE REVIEW

**Comment by hammar** [comment]
Noted an issue with enums in ADT Explorer: they apparently cannot be unset, once set, and the ADT Explorer GUI defaults to showing an enum field as filled with one (the first?) value in the enum schema. I.e., it's really easy in the GUI to mess up enum-using twins and not be able to undo one's changes. So, for those cases where we use enums: should we provide a safe "unset" option in the enum?

**Comment by hammar** [comment]
> Noted an issue with enums in ADT Explorer: they apparently cannot be unset, once set, and the ADT Explorer GUI defaults to showing an enum field as filled with one (the first?) value in the enum schema. I.e., it's really easy in the GUI to mess up enum-using twins and not be able to undo one's changes. So, for those cases where we use enums: should we provide a safe "unset" option in the enum?

As a matter of fact, ADT Explorer does not seem to play nice with enums GUI-wise even with a solution such as the one proposed above. Perhaps best to update it to do so, rather than implement workarounds in the models.

File: Ontology/Capability/Capability.json [hunk 1]
```diff
@@ -14,6 +14,124 @@
...
+    {
+      "@type": "Property",
+      "description": {
+        "en": "The (optional) capability position relative to the system or space that it interacts with. E.g., a water temperature sensor can be positioned on the supply or return flow, air humidity can be measured entering or leaving a system, etc."
+      },
+      "displayName": {
+        "en": "position"
+      },
+      "name": "position",
+      "dtmi:dtdl:property:schema;2": {
```
Comment (alinamstanciu) [inline Ontology/Capability/Capability.json:83]: I see more values in the same excel, you just missed them or don't agree with the list? Thanks

File: Ontology/Capability/Sensor/ConductivitySensor.json [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {
-  "@id": "dtmi:digitaltwins:rec_3_3:device:MotionSensor;1",
+  "@id": "dtmi:digitaltwins:rec_3_3:device:ConductivitySensor;1",
```
Comment (alinamstanciu) [inline Ontology/Capability/Sensor/ConductivitySensor.json:2]: Are we removing MotionSensor?

File: Ontology/Capability/State.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
+{
```
Comment (alinamstanciu) [inline Ontology/Capability/State.json:1]: Are we also adding a property as enum for all states listed in the same excel?

**Review by alinamstanciu — APPROVED** [review]
LGTM with minor changes.
Karl, this is a partial change, we still need more changes in order to represent the Optio3-Alina sensors, setpoints, etc. 
Thanks Karl!

**Comment by alinamstanciu** [comment]
@hammar , 
For ADT-Explorer I agree with this " As a matter of fact, ADT Explorer does not seem to play nice with enums GUI-wise even with a solution such as the one proposed above. Perhaps best to update it to do so, rather than implement workarounds in the models."
@cschormann, could you please make sure you address ADT Explorer issues with enums? See above.

Thanks

**Comment by rszcodronski** [comment]
@hammar @alinamstanciu 
I have just merged a bunch of changes into Willow Ontology to align with the spreadsheet I prepared on this Capability topic:
https://github.com/WillowInc/opendigitaltwins-building/tree/main/Ontology/Willow/Capability

There are a few differences from what I see here and I've fully populated the Capability enums for position, phenomenon, etc.

**Comment by hammar** [comment]
> @hammar ,
> For ADT-Explorer I agree with this " As a matter of fact, ADT Explorer does not seem to play nice with enums GUI-wise even with a solution such as the one proposed above. Perhaps best to update it to do so, rather than implement workarounds in the models."
> Could you please show repro steps, not sure I understand correctly.
> 
> @cschormann, could you please make sure you address ADT Explorer issues with enums? See above.
> 
> Thanks

I'm uploading five pictures illustrating different parts of the process/problem:

* [The enum schema](https://user-images.githubusercontent.com/68027/112294517-360f8500-8c93-11eb-803e-7bf341bea6ab.png)
* [Twin that uses this schema](https://user-images.githubusercontent.com/68027/112294514-3576ee80-8c93-11eb-975e-e1bdcfcfebda.png). Note that "substance" and "position" fields are seemingly prepopulated with the first enum value for the respective property, in spite of these properties not actually being set on the twin yet.
* [Result of saving the twin with that default value prepopulated from the schema](https://user-images.githubusercontent.com/68027/112294521-360f8500-8c93-11eb-8c3c-b4a1a3254e23.png). Note the lack of patch.
* [Saving a value that does not exist in the enum definition](https://user-images.githubusercontent.com/68027/112294522-360f8500-8c93-11eb-80fd-e24ec88305e9.png). Note that this actually saves the default (first listed value), rather than throw an error as expected.
* [Saving another value that is consistent with the enum schema definition](https://user-images.githubusercontent.com/68027/112294524-36a81b80-8c93-11eb-9987-7c3267fa6401.png). In this case I manually typed in "Steam", an allowed value, and saved. Note that this actually saves the default (first listed value) rather than the one I entered.

File: Ontology/Capability/Sensor/ConductivitySensor.json (hunk 2 above)
Comment (hammar) [reply Ontology/Capability/Sensor/ConductivitySensor.json:2]: No, it has been moved down into the "BinarySensor" hierarchy, which simplifies typing of the "lastValue" property for all of those sensors that return binary values. If you think this is an inappropriate clustering, it is not a dealbreaker for me to revert.

File: Ontology/Capability/Capability.json [hunk 4]
```diff
@@ -14,6 +14,124 @@
...
     },
+    {
+      "@type": "Property",
+      "description": {
+        "en": "The (optional) substance that a capability measures or interacts with; e.g., a TemperatureSensor can measure air temperature or water temperature, a Flow Demand Setpoint can be configured to control steam demand or chilled water demand, etc."
+      },
+      "displayName": {
+        "en": "substance"
+      },
+      "name": "substance",
```
[the remark this answers is not in the document]
Comment (hammar) [reply Ontology/Capability/Capability.json:25]: I based this list on the Optio3 list, which I took to be the target list for the first iteration -- but changing to the list provided by Rick is not a problem and I will do so shortly (in an hour or so) and update this PR. I'll also rename substance to phenomenon.

File: Ontology/Capability/Capability.json (hunk 1 above)
Comment (hammar) [reply Ontology/Capability/Capability.json:83]: See previous comment.

File: Ontology/Capability/State.json (hunk 3 above)
Comment (hammar) [reply Ontology/Capability/State.json:1]: Don't really understand the question, but my understanding is that we're still in a bit of flux re. the design of the state classes, so this is mostly a placeholder for now.

**Comment by hammar** [comment]
> @hammar @alinamstanciu
> I have just merged a bunch of changes into Willow Ontology to align with the spreadsheet I prepared on this Capability topic:
> https://github.com/WillowInc/opendigitaltwins-building/tree/main/Ontology/Willow/Capability
> 
> There are a few differences from what I see here and I've fully populated the Capability enums for position, phenomenon, etc.

Posting in the same response here as I did to Alina's other comment (as I'm not sure if you're pinged on that):

> I based this list on the Optio3 list, which I took to be the target list for the first iteration -- but changing to the list provided by Rick is not a problem and I will do so shortly (in an hour or so) and update this PR. I'll also rename substance to phenomenon.

**Comment by hammar** [comment]
The phenomena and position lists are now updated, I hope this is more in line with expectations.

Note that since RDF, JSON, and JSON-LD do not care about key ordering, enums can sometimes be serialized in a semi-random order when the models are generated. This is obviously less than ideal and I will try to dig into the relevant libraries to see if I can mitigate the effect.

**Comment by alinamstanciu** [comment]
> > @hammar ,
> > For ADT-Explorer I agree with this " As a matter of fact, ADT Explorer does not seem to play nice with enums GUI-wise even with a solution such as the one proposed above. Perhaps best to update it to do so, rather than implement workarounds in the models."
> > Could you please show repro steps, not sure I understand correctly.
> > @cschormann, could you please make sure you address ADT Explorer issues with enums? See above.
> > Thanks
> 
> I'm uploading five pictures illustrating different parts of the process/problem:
> 
> * [The enum schema](https://user-images.githubusercontent.com/68027/112294517-360f8500-8c93-11eb-803e-7bf341bea6ab.png)
> * [Twin that uses this schema](https://user-images.githubusercontent.com/68027/112294514-3576ee80-8c93-11eb-975e-e1bdcfcfebda.png). Note that "substance" and "position" fields are seemingly prepopulated with the first enum value for the respective property, in spite of these properties not actually being set on the twin yet.
> * [Result of saving the twin with that default value prepopulated from the schema](https://user-images.githubusercontent.com/68027/112294521-360f8500-8c93-11eb-8c3c-b4a1a3254e23.png). Note the lack of patch.
> * [Saving a value that does not exist in the enum definition](https://user-images.githubusercontent.com/68027/112294522-360f8500-8c93-11eb-80fd-e24ec88305e9.png). Note that this actually saves the default (first listed value), rather than throw an error as expected.
> * [Saving another value that is consistent with the enum schema definition](https://user-images.githubusercontent.com/68027/112294524-36a81b80-8c93-11eb-9987-7c3267fa6401.png). In this case I manually typed in "Steam", an allowed value, and saved. Note that this actually saves the default (first listed value) rather than the one I entered.

Thanks Karl!!!!

**Comment by alinamstanciu** [comment]
> > @hammar @alinamstanciu
> > I have just merged a bunch of changes into Willow Ontology to align with the spreadsheet I prepared on this Capability topic:
> > https://github.com/WillowInc/opendigitaltwins-building/tree/main/Ontology/Willow/Capability
> > There are a few differences from what I see here and I've fully populated the Capability enums for position, phenomenon, etc.
> 
> Posting in the same response here as I did to Alina's other comment (as I'm not sure if you're pinged on that):
> 
> > I based this list on the Optio3 list, which I took to be the target list for the first iteration -- but changing to the list provided by Rick is not a problem and I will do so shortly (in an hour or so) and update this PR. I'll also rename substance to phenomenon.

@hammar , are we eventually going the way Rick did it ? https://github.com/WillowInc/opendigitaltwins-building/tree/main/Ontology/Willow/Capability/Sensor
Thanks

**Comment by rszcodronski** [comment]
Hi all,
I just completed another large push within Willow Capabilities as I've iterated on the ideas posted above by @hammar around clustering, etc. Here's a link to Willow Capability:
https://github.com/WillowInc/opendigitaltwins-building/tree/main/Ontology/Willow/Capability

Key updates:
* Introduced subclasses to Sensor/Setpoint to cluster on the following: Quantity, Detection, Status (still needs work)
* The above allows us to move the lastValue to the grouping since they all have the same schema (double, boolean) below
* I chose to use these terms over the even more generic Numeric/Double/Boolean because I think they are more intuitive and we don't need to introduce an intermediate clustering of different Boolean sensors (i.e. Detection and Status) at this time
* Unlike Sensor/Setpoint clustering, I do think it makes sense to cluster State/Actuator on the terms Binary(Boolean?), Multi-State, and PercentUnity given the nature of how these clusters come together.
* The Willow ontology is representative of my current thinking. If/once the concepts are taken up by this ontology, we will remove duplicate properties (i.e. lastValue) and just extend this ontology.

**Comment by hammar** [comment]
@rszcodronski that looks very promising. I'm a bit under-the-weather at the moment (hopefully just a cold..) and not operating at full capacity; so a more thorough analysis from me may take a day or two more. In the interim I've asked @erikoskarwallin to check this design versus the Idun platform and needs, and see if we can get the REC consortium on board. Personally I think (again, subject to actually looking a little deeper, once my head is working properly) that I'm on board.

Cheers,

Karl

**Review by alinamstanciu — APPROVED** [review]
LGTM

**Comment by hammar** [comment]
Note that this contains the REC view on quantitysensors, including semantic typing where applicable, etc. I put together this list with the goal to fulfil the Optio3 requirements, but it is not a 1:1-mapping of every exact concept against Optio3 and it is shorter and structured somewhat differently (I'd argue more concise/usable, but I recognize that opinions may diverge on this) than the Willow list. 

Anyone else who has input on this, or should we merge and proceed?

**Pull request merged**